### PR TITLE
feat: add support for macOS 13 Ventura

### PR DESCRIPTION
**Pull request by ShlomoCode**
## Describe your changes
Added support for macOS 13 Ventura.

## Related issue number(s) and link(s)
closes https://github.com/ejbills/DockDoor/issues/211

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If this change affects core functionality, I have added a description highlighting the changes
- [x] I have followed conventional commit guidelines

## Core Functionality Changes

No change is expected for macOS 14 and above, for older systems support does not include a number of small things such as SmoothGradient ~~or the option to cancel in warning about overlapping elements in the settings~~.

### CODE REVIEW

File: DockDoor/Components/Marquee.swift
Comment (ejbills) [inline DockDoor/Components/Marquee.swift:66]: Let's keep these as underscores, the variable is not used.

File: DockDoor/Utilities/WindowUtil.swift
Comment (ejbills) [inline DockDoor/Utilities/WindowUtil.swift:105]: This is kind of important, why is that removed?

File: DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift
Comment (ejbills) [inline DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift:1]: Does the window switcher work with this? This was the main reason DD is min macOS 14, the observable property allows us to reflect the changes in index when using the switcher. I'm not sure if @Published properties are a slot-in replacement - however I may be wrong.

File: DockDoor/Utilities/WindowUtil.swift
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift:105]: `'captureResolution' is only available in macOS 14.0 or newer`
So we need to condition it to avoid compiler error...
Maybe I'll separate it into a separate "captureWindowModern" function and mark it all as available(macOS 14.0, *), that would be nicer

**Comment by ejbills** [comment]
> option to cancel in warning about overlapping elements in the settings

Is this really not a thing in previous versions? I think we can make the message util take another approach if it's not available, similar to the window manager legacy func you added. 

Cool PR, thank you! Love it!

File: DockDoor/Components/Marquee.swift
Comment (ShlomoCode) [reply DockDoor/Components/Marquee.swift:66]: I just removed the second parameter as it is not supported in macOS 13, i did not change it to a named parameter

File: DockDoor/Utilities/WindowUtil.swift
Comment (ejbills) [reply DockDoor/Utilities/WindowUtil.swift:105]: Yes, keep in mind the option of specifying a capture resolution has existed for a while. I'm not exactly sure how to set it on the legacy screen API though... but it is possible.

**Comment by ShlomoCode** [comment]
The message still appears, only the "cancel" button will not work (we do need to hide it), but the main goal, to draw the user's attention to the overlap, has been achieved. How can I get the oldValue in macOS 13?

File: DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift
Comment (ShlomoCode) [reply DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift:1]: It seems to work fine, although I'm actually not very familiar with the windows switcher. Is there anything special I should look out for?

Uploading CleanShot 2024-08-23 at 04.15.05.mp4…

**Comment by ejbills** [comment]
> How can I get the oldValue in macOS 13?

Store the value in a state var within the settings view.

File: DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift
Comment (ejbills) [reply DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift:1]: I'll try it out in a bit

**Comment by ShlomoCode** [comment]
Thanks, done.

**Comment by ShlomoCode** [comment]
@ejbills One thing that is strange to me is that the image appears to be quarter size, can you figure out why this is happening? 
CleanShot 2024-08-23 at 05 03 11@2x

You can restore it in current macOS by changing this:
```swift
let image: CGImage = if #available(macOS 14.0, *) {
    try await captureWindowImageModern(window: window, width: width, height: height)
} else {
    try await captureWindowImageLegacy(for: window, width: width, height: height)
}
```

From 14 to 19 for example...

File: DockDoor/Utilities/WindowUtil.swift
Comment (ShlomoCode) [reply DockDoor/Utilities/WindowUtil.swift:105]: Yes `.bestResolution`, but I already define it

**Comment by ShlomoCode** [comment]
Tested with macOS 13.0 VM, looks fine
I think it's ready

**Comment by ejbills** [comment]
> Tested with macOS 13.0 VM, looks fine I think it's ready

The window switcher does not function properly, pressing tab now does nothing as the @Observable was responsible for making the view react to the index changes.

**Comment by ejbills** [comment]
@ShlomoCode I am working on integrating the observed object, please don't push anything

**Comment by ejbills** [comment]
@ShlomoCode can you give me write permissions for your fork?

**Review by ejbills — CHANGES_REQUESTED** [review]
https://github.com/ejbills/DockDoor/pull/267#issuecomment-2307416825

**Comment by illavoluntas** [comment]
Once this PR is merged, does it mean DockDoor will be fully compatible with Ventura at the next release?

Does the API used for Ventura work on Monterey? Or is it 13.0+?

**Comment by ShlomoCode** [comment]
> Once this PR is merged, does it mean DockDoor will be fully compatible with Ventura at the next release?
> 

Yes

> Does the API used for Ventura work on Monterey? Or is it 13.0+?

I can adapted quite easily to macOS 12.3, the question is how much demand is there for it?

**Comment by ShlomoCode** [comment]
@ejbills Sorry for the late reply, you should now be able to push. Pay attention to do a rebase pull first  (`git pull --rebase`)

**Comment by illavoluntas** [comment]
> I can adapted quite easily to macOS 12.3, the question is how much demand is there for it?

There will always be demand for older system compatibility. As of today, I think some people still use 10.15 Catalina, it might be the oldest most demanded version (macOS 11 Big Sur abandoned a lot of Macs).

**Comment by ShlomoCode** [comment]
OK, let's merge Ventura first.

**Comment by ShlomoCode** [comment]
@ejbills I see that fixing the broken onChange currIndex is a small change, so I just pushed it

**Comment by ejbills** [comment]
> @ejbills Sorry for the late reply, you should now be able to push. Pay attention to do a rebase pull first (`git pull --rebase`)



> @ejbills Sorry for the late reply, you should now be able to push. Pay attention to do a rebase pull first (`git pull --rebase`)

My local changes actually get rid of the coordinator being a singleton. I rewrote the logic to ensure it properly makes use of the ObservableObject pattern for passing data to child views. So, the onChange won't be necessary. I will push later today, thanks!

**Comment by ShlomoCode** [comment]
It sounds more appropriate for a separate refactor commit

**Comment by ejbills** [comment]
> It sounds more appropriate for a separate refactor commit

 I don't think a separate refactor commit is necessary in this case. The change to an ObservableObject pattern is directly related to the PR's purpose of adding macOS Ventura support. It's a relatively small change, impacting less than ~100 lines of code. I'll be pushing these changes soon as part of this PR, as I am away from my computer at the moment. 

Further refactoring is on my todo list, and it will not be in this PR. Hope that addresses your concerns.

**Comment by ShlomoCode** [comment]
@ejbills You dropped some of my code...
CleanShot 2024-08-25 at 03 49 37@2x
CleanShot 2024-08-25 at 03 49 29@2x

When working on PR together, it is advisable to wait for a mutual review

**Pull request merged**
